### ISSUE TITLE
Custom Primary keys not listed in `persistUniqueKeys`

### ISSUE THREAD

**Issue by tysonzero** [yesodweb/persistent#1090]
When specifying a custom `Primary` in persistent-template, the field is required to be unique. However this field is not listed in `persistUniqueKeys`, so various operations involving uniqueness ignore it.

**Comment by parsonsmatt** [comment]
Yeah, this is frustrating. A lot of the `upsert` machinery is complicated due to this problem.

Let me lay out the cases:

```
Normal
    name Text
    age Int

OneUnique
    name Text
    age Int
    
    UniqueName name

UniquePrimary
    name Text
    age Int
    
    Primary name
    UniqueAge age

JustPrimary
    name Text
    age Int
    
    Primary name
```

`Normal` has an empty `data Unique Normal` definition - there's nothing there. `OneUnique` has a definition like `data Unique OneUnique = UniqueName Text`. `JustPrimary` *does not* have a `Unique JustPrimary` constructor at all, and then `UniquePrimary` *does* have a `UniqueAge` constructor, but there's no accounting for the `Primary name` declaration at all.

A lot of the assumptions in the code are based on the idea that you can do `PersistEntity record => record -> [Unique record]`. But this isn't true for `Normal` - you'd need `PersistEntity record => Entity record -> [Unique record]` to get that.

I think simply including a `Primary` declaration in the `entityUniques` would work. It'd be a breaking change, since it would affect a lot of the code generation, but only in ways that allow previously illegal states.

**Comment by tysonzero** [comment]
> A lot of the assumptions in the code are based on the idea that you can do PersistEntity record => record -> [Unique record]. But this isn't true for Normal - you'd need PersistEntity record => Entity record -> [Unique record] to get that.

It seems like the proposed entity changes in #995 would workaround this problem right?

**Comment by parsonsmatt** [comment]
It would, yes. I need to devote some time and energy towards that. But this is a quicker fix, I think!

**Comment by neilmayhew** [comment]
I ran into this problem. I want to be able to use `putMany` on an entity that has a custom, compound PK, but the TH defines `persistUniqueKeys` to be `[]`.

My workaround is to include both a `Primary` and a `Unique` declaration in the QQ schema, and to ignore the migration warning telling me to add a uniqueness constraint to the DB schema:

```
Message
    subject   Text
    body      Text
    sent      Bool
    contact   ContactId
    Primary subject contact
    UniqueMessageSubjectContact subject contact
```

I have a few questions, though:

1. Is this workaround a reasonable thing to do?

2. How long do you think it will be before there's a proper fix in the library itself?

3. Couldn't the TH just assume an implied `Unique` matching any `Primary`, since that's how the DB works anyway?

The other solution is to use `repsertMany`. However, this requires using `keyFromRecordM` to get a `Maybe keyFromRecord` and that's rather ugly, because you either have to get it ahead of time and save it somewhere, or deal with it being `Nothing` at runtime.

**Comment by malteneuss** [comment]
I also stumbled on this rough edge here. Thanks @neilmayhew  for the workaround. We should probably mention this workaround in the docs (until a fix is in place; i could do that if some of the maintainers thinks this is a good idea) given that a solution hasn't come up in 4 years

**Comment by parsonsmatt** [comment]
oof, yeah, we should fix this. patch would be super welcome!!
